### PR TITLE
Lib: SMF: Add initial transitions to HSMs and add smf_set_handled()

### LINKED ISSUE #66341: SMF ancestor states do not handle transition to self correctly
### Discussed in https://github.com/zephyrproject-rtos/zephyr/discussions/66112

<div type='discussions-op-text'>

<sup>Originally posted by **glenn-andrews** December  4, 2023</sup>
I think there might be an issue with the SMF handling of transitions to self in hierarchical state machines. 

A transition to self is when the run action of a state calls `smf_set_state()` with a target of itself. The exit and entry actions of the state should be called in that order.

The issue is when a child state triggers an event that causes a transition to self in a parent state. This should trigger a transition from the child state to the parent state and the following actions: child exit, parent exit, parent entry.

When ancestor run actions are executed, the context doesn't change, so a transition to self in a parent state looks like a transition from the child to the parent, not from parent to itself, and doesn't fire the exit action of the parent state. The result is: child exit, parent entry.

Is there any way to fix this behavior?

Example from the PSiCC2 book by Miro Samek:
image

When in state s11, receiving event A will propagate up to s1, which will initiate a transition to self. You should get s11 exit, s1 exit and s1 entry. (you should also then get an initial transition to s11, with s11 entry, but the SMF doesn't support initial transitions - yet)</div>

### PR DESCRIPTION
**Pull request by glenn-andrews**
Brings SMF framework closer into alignment with accepted Hierarchical State Machine operation by:
1. Allowing 'programming by difference' by having some child states handle events and prevent propagation up to the parent run actions while others propagate events up to a common handler in a parent state.
2. Optionally allow initial transitions within a parent state to determine the most nested child state to transition to.
3. Adding a test case for `CONFIG_SMF_INITIAL_TRANSITION` and `smf_set_handled()`
4. Updating documentation for the new API (and fixing some references)

There was discussion in https://github.com/zephyrproject-rtos/zephyr/issues/55344 about not making the initial transition a Kconfig option, but I'm not sure of any way else of doing it without permanently adding a pointer to each `smf_state` entry, which is a problem for resource-constrained devices.

This does not fix https://github.com/zephyrproject-rtos/zephyr/issues/66341 but documentation has been updated to warn users of the issue.

### CODE REVIEW

File: lib/smf/smf.c [hunk 1]
```diff
@@ -18,14 +18,12 @@ struct internal_ctx {
...
 };
 
-static bool share_paren(const struct smf_state *test_state,
-			const struct smf_state *target_state)
+static bool share_paren(const struct smf_state *test_state, const struct smf_state *target_state)
 {
-	for (const struct smf_state *state = test_state;
-	     state != NULL;
-	     state = state->parent) {
+	for (const struct smf_state *state = test_state; state != NULL; state = state->parent) {
```
Comment (sambhurst) [inline lib/smf/smf.c:26]: please run clang-format

Comment (glenn-andrews) [reply lib/smf/smf.c:26]: I did, and that was the result, plus messing up all the struct definitions.

Is there some magic I need to invoke on Windows other than `clang_format -i smf.c` as recommended by the contribution guidelines?

Comment (glenn-andrews) [reply lib/smf/smf.c]: On recommendation of the Discord, I've reverted all clang-format changes.

Comment (glenn-andrews) [reply lib/smf/smf.c]: > please run clang-format

Is the update satisfactory?

Comment (sambhurst) [reply lib/smf/smf.c]: Looks good to me

**Review by sambhurst — DISMISSED** [review]

File: include/zephyr/smf.h [hunk 2]
```diff
@@ -136,6 +163,15 @@ void smf_set_state(struct smf_ctx *ctx, const struct smf_state *new_state);
...
 void smf_set_terminate(struct smf_ctx *ctx, int32_t val);
 
+/**
+ * @brief Tell the SMF to stop propagating the event to ancestors. This allows
+ *        HSMs to implement 'programming by difference' where substates can
+ *        handle events on their own or propagate up to a common handler.
+ *
+ * @param ctx  State machine context
+ */
+void smf_set_handled(struct smf_ctx *ctx);
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply include/zephyr/smf.h:173]: I have a rather large sample that implements the PSiCC2 Section 2.3.15 demo app with all levels of transitions. It's about 11K of source, given the large number of states the demo app encompasses.

[the remark this answers is not in the document]
Comment (keith-zephyr) [reply include/zephyr/smf.h:173]: That's probably too large.  I was thinking we had existing samples that use the SMF, but there are not.

However, there are some good unit tests for the SMF [here](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/lib/smf/src).

Please extend one of the existing tests or create a new one that enables the `CONFIG_SMF_INITIAL_TRANSITION` option.  The ensures that Zephyr's CI will validate that this new feature.

[the remark this answers is not in the document]
Comment (glenn-andrews) [reply include/zephyr/smf.h:173]: Done. I've never used the unit tests before so feedback is appreciated.

[the remark this answers is not in the document]
Comment (glenn-andrews) [reply include/zephyr/smf.h:173]: > That's probably too large. I was thinking we had existing samples that use the SMF, but there are not.

Would it be worth me submitting my demo app as a separate sample for SMF? I can probably simplify it a bit.

[the remark this answers is not in the document]
Comment (kartben) [reply include/zephyr/smf.h:173]: > > That's probably too large. I was thinking we had existing samples that use the SMF, but there are not.
> 
> Would it be worth me submitting my demo app as a separate sample for SMF? I can probably simplify it a bit.

Yes, please!

File: doc/services/smf/index.rst [hunk 3]
```diff
@@ -39,6 +39,10 @@ By default, a state can have no ancestor states, resulting in a flat state
 machine. But to enable the creation of a hierarchical state machine, the
 :kconfig:option:`CONFIG_SMF_ANCESTOR_SUPPORT` option must be enabled.
 
+By default, the hierarchical state machine does not support initial transitions
+to child states on entering a superstate. To enable them the
+:kconfig:option:`CONFIG_SMF_INITIAL_TRANSITION` option must be enabled.
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply doc/services/smf/index.rst:44]: I tried my hand at a graphviz statechart. Couldn't wok out how to get an initial transition from the outside to STATE_S0. Let me know if it's OK as is.

File: doc/services/smf/index.rst [hunk 4]
```diff
@@ -71,6 +75,21 @@ And this example creates three hierarchical states::
...
+   /* Forward declaration of state table */
+   const struct smf_state demo_states[];
+
+   const struct smf_state demo_states[] = {
+      [S0] = SMF_CREATE_STATE(s0_entry, s0_run, s0_exit, NULL, demo_states[S2]),
+      [S1] = SMF_CREATE_STATE(s1_entry, s1_run, s1_exit, demo_states[S0], NULL),
+      [S2] = SMF_CREATE_STATE(s2_entry, s2_run, s2_exit, demo_states[S0], NULL)
+   };
+
+
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply doc/services/smf/index.rst:110]: See new note.

File: doc/services/smf/index.rst [hunk 5]
```diff
@@ -317,7 +343,11 @@ When designing hierarchical state machines, the following should be considered:
    example, the s1_exit function is called before the parent_exit function
    is called.
  - The parent_run function only executes if the child_run function returns
-   without transitioning to another state, ie. calling smf_set_state.
+   without transitioning to another state, ie. calling smf_set_state, or if
+   ``smf_set_handled`` is called within the child_run function.
+ - When a parent state intitiates a transition to self, the parents's exit
+   action is not called, e.g. instead of child_exit, parent_exit, parent_entry
+   it performs child_exit, parent_entry
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply doc/services/smf/index.rst:374]: Do we want to be testing to confirm a bug exists? I don't want to see this codified as "the way things work" and add additional effort to fixing it by having to fix the test.

File: tests/lib/smf/src/test_lib_initial_transitions_smf.c [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+		      "Incorrect test value index for parent entry termination");
+	zassert_equal(test_obj.transition_bits, test_value[test_obj.tv_idx],
+		      "Final parent entry termination state not reached");
+
+	/* C) Test termination in parent run action */
+
+	test_obj.transition_bits = 0;
+	test_obj.first_time = 1;
+	test_obj.terminate = PARENT_RUN;
+	smf_set_initial((struct smf_ctx *)&test_obj, &test_states[STATE_A]);
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply tests/lib/smf/src/test_lib_initial_transitions_smf.c]: Oops. Fixed.

File: doc/services/smf/index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply doc/services/smf/index.rst:44]: The graphviz appears to be having trouble. Can you see the issue? I've never used it before.

File: doc/services/smf/index.rst [hunk 7]
```diff
@@ -79,6 +117,13 @@ To transition from one state to another, the ``smf_set_state`` function is
 used and it has the following prototype:
 ``void smf_set_state(smf_ctx *ctx, smf_state *state)``
 
+**NOTE:** If :kconfig:option:`CONFIG_SMF_INITIAL_TRANSITION` is not set,
```
Comment (kartben) [inline doc/services/smf/index.rst:120]: There is a `.. note::` Sphinx directive that would probably be appropriate here and in other places with `**NOTE**`s

File: doc/services/smf/index.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (keith-zephyr) [reply doc/services/smf/index.rst:375]: The documentation here defines an API contract:  A parent state initiating a transition to self does not call the parent_exit.  I didn't consider this a bug.
So I wondering if this condition could explicitly be tested.

File: doc/services/smf/index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (kartben) [reply doc/services/smf/index.rst:44]: > The graphviz appears to be having trouble. Can you see the issue? I've never used it before.

you fixed it, yes? FWIW I couldn't find a solution for drawing the transition symbolizing that S0 is the entry state :-/

File: doc/services/smf/index.rst [hunk 8]
```diff
@@ -71,6 +78,37 @@ And this example creates three hierarchical states::
    };
 
 
+This example creates three hierarchical states with an initial transition
+from parent state S0 to child state S2::
```
Comment (keith-zephyr) [inline doc/services/smf/index.rst:82]: This section should be added following the Hierarchical State Machine Example.  It should also include the code showing state transitions.

As is, I don't think this diagram is helpful to the user as it's demonstrating too trivial a case.

But I'm okay with you dropping this example/diagram from this PR and creating a follow-up PR to make the improvements here.

Comment (keith-zephyr) [reply doc/services/smf/index.rst:82]: I think a good goal here would be to have a diagram that matches the state transitions from the test.

File: doc/services/smf/index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (keith-zephyr) [reply doc/services/smf/index.rst:44]: This gets close to my understanding the transitions for the state machine defined by your test:
```
.. graphviz::
   :caption: demo_states with initial transition to S2

   digraph smf_hierarchical_initial {
      compound=true;
      node [style = rounded];
      smf_set_initial [shape=plaintext];
      ab_init_state [shape = point];
      STATE_A [shape = box];
      STATE_B [shape = box];
      STATE_C [shape = box];
      STATE_D [shape = box];

      subgraph cluster_ab {
         label = "PARENT_AB";
         style = rounded;
         ab_init_state -> STATE_A;
         STATE_A -> STATE_B;
      }

      subgraph cluster_c {
         label = "PARENT_C";
         style = rounded;
	 STATE_C -> STATE_C
      }

      smf_set_initial -> STATE_A [lhead=cluster_ab]
      STATE_B -> STATE_C
      STATE_C -> STATE_D
   }
```

File: doc/services/smf/index.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply doc/services/smf/index.rst:375]: It's considered a bug here: https://github.com/zephyrproject-rtos/zephyr/issues/66341

File: doc/services/smf/index.rst [hunk 9]
```diff
@@ -91,6 +110,13 @@ some application dependent way. An application should cease calling
 smf_run_state if it returns a non-zero value. The function has the following
 prototype: ``int32_t smf_run_state(smf_ctx *ctx)``
 
+Preventing Parent Run Actions
+=============================
+
+Calling ``smf_set_handled`` prevents calling the run action of parent states.
+It is not required to call ``smf_set_handled`` if the state calls
+``smf_set_state``.
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply doc/services/smf/index.rst:144]: Do you want me to refactor the entire document, or just the parts I've touched?

**Comment by glenn-andrews** [comment]
Anything else I should do?

**Review by kartben — CHANGES_REQUESTED** [review]
@glenn-andrews thanks much for your work on this! I've pushed a couple commits to your branch adding what was missing for the functions to properly render as links.

The one thing that I think needs to be changed is that you have *all* your changes into just one commit (which description now seems outdated/incomplete anyway?), which is not ideal. It would be great if you could split in several commits so that the additions of new Kconfigs and APIs are not mixed with what are "simple" documentation enhancements.

Thanks!

**Comment by glenn-andrews** [comment]
> It would be great if you could split in several commits so that the additions of new Kconfigs and APIs are not mixed with what are "simple" documentation enhancements.

I'm not entirely sure how such a split would work. We've got about 8 instances where I replaced double-backticks with the appropriate sphinx directives, all the rest is directly relevant to the API being added. I could revert the changes to the double-backticks, then artificially add them again, but that seems silly.

Can you give me a clearer idea of what is required? I feel I'm missing the intent of this request.

**Comment by glenn-andrews** [comment]
> I've pushed a couple commits to your branch adding what was missing for the functions to properly render as links.

I'm terribly sorry, but I think I managed to nuke your changes by not pulling them before I amended my commit message. I fixed the `smf_terminate` one, but can't remember where the doxygen one goes.

**Comment by glenn-andrews** [comment]
How's this looking now? Any more action items for me?

File: doc/services/smf/index.rst (hunk 8 above)
Comment (glenn-andrews) [reply doc/services/smf/index.rst:82]: Moved diagram to end of document.

**Comment by glenn-andrews** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/66753#discussion_r1510029400

Updated. Does the new wording seem okay?

**Review by keith-zephyr — APPROVED** [review]

**Review by kartben — APPROVED** [review]
Thanks a lot for all the hard work, @glenn-andrews!

**Comment by glenn-andrews** [comment]
Thank you all so much for reviewing this!

**Review by sambhurst — APPROVED** [review]

File: doc/services/smf/index.rst [hunk 10]
```diff
@@ -466,3 +504,48 @@ Code::
...
+
+Hierarchical State Machine Example With Initial Transitions
+===========================================================
+
+:zephyr_file:`tests/lib/smf/src/test_lib_initial_transitions_smf.c` defines
+a state machine for testing initial transitions and :c:func:`smf_set_handled`.
+The statechart for this test is below.
+
+.. graphviz::
+   :caption: Test state machine for initial trnasitions and ``smf_set_handled``
```
Comment (bryceschober) [inline doc/services/smf/index.rst:516]: Typo "trnasitions"

**Pull request merged**